app/models/filter.rb

```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+
+class Filter
+  include ActiveModel::Model
+
+  QUERY_PARAMS = [:url_pattern, :keyword_pattern, { link_types: [] }].freeze
+  PATTERN_REGEXP =
+    %r{\A\^?((\[[\\\-/:.\w]+\])|([\\\-/:.\w]+)|(\(([\\\-/:.\w]+)\|?([\\\-/:.\w]+)\)))[+?*]?(\{\d+(,\d?)?\})?\$?\Z}
+    .freeze
+
+  attr_accessor :keyword_pattern, :url_pattern, :link_types
+
+  validates :keyword_pattern, format: { with: PATTERN_REGEXP }, allow_blank: true
+  validates :url_pattern, format: { with: PATTERN_REGEXP }, allow_blank: true
+  validate :link_types_exist?
+
+  def self.from_params(params)
```

Comment:
How about adding unit test for this method and `result_link_filter_exist?` method?